File: Cores/ExtremumScalaCore/src/main/scala/OSOL/Extremum/Core/Scala/Optimization/Algorithm.scala
package OSOL.Extremum.Core.Scala.Optimization

import scala.util.control.Breaks._
import OSOL.Extremum.Core.Scala.Optimization.Nodes.GeneralNode

final class Algorithm[Base, FuncType, V <: Optimizable[Base, FuncType]] private
(nodes: Seq[GeneralNode[Base, FuncType, V]], transitionMatrix: Seq[(Int, Option[Int], Int)]) {

  private var state: State[Base, FuncType, V] = null

  private var currentNode: GeneralNode[Base, FuncType, V] = null

  def initialize(f: Map[String, FuncType] => FuncType, area: Area): Unit = nodes.foreach(_.initialize(f, area, state))

  def work(f: Map[String, FuncType] => FuncType, area: Area): V = {
    initialize(f, area)
    while (true) {
      currentNode.process(f, area, state)
      val currentConditionValue = currentNode.getCurrentCondition(f, area, state)
      val nextNode = transitionMatrix.find { case (currentId, condition, _) =>
        currentId == currentNode.nodeId && condition == currentConditionValue }
      if (nextNode.isDefined) { currentNode = nodes.find(_.nodeId == nextNode.get._3).get }
      else break
    }
    state.elements.minBy(_.getPerformance(f))
  }

}

File: Cores/OSOL.Extremum.Core.DotNet/Optimization/Algorithm.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Newtonsoft.Json.Linq;
using OSOL.Extremum.Core.DotNet.Optimization.Nodes;

namespace OSOL.Extremum.Core.DotNet.Optimization
{
    public class Algorithm<TBase, TFuncType, TV> where TV: class, IOptimizable<TBase, TFuncType>
    {
        public State<TBase, TFuncType, TV> State = new State<TBase, TFuncType, TV>();
        public GeneralNode<TBase, TFuncType, TV> CurrentNode = null;

        public GeneralNode<TBase, TFuncType, TV>[] Nodes;
        public Tuple<int, int?, int>[] TransitionMatrix;

        public void Initialize(Func<Dictionary<string, TFuncType>, TFuncType> f, Area area)
        {
            foreach(var n in Nodes)
                n.Initialize(f, area, State);
        }

        public TV Work(Func<Dictionary<string, TFuncType>, TFuncType> f, Area area)
        {
            Initialize(f, area);
            bool continueOrNot = true;
            while (continueOrNot)
            {
                CurrentNode.Process(f, area, State);
                var currentConditionValue = CurrentNode.GetCurrentCondition(f, area, State);
                var nextNodes = TransitionMatrix
                    .Where(rule => rule.Item1 == CurrentNode.NodeId && rule.Item2 == currentConditionValue)
                    .Select(rule => (double) rule.Item3).ToArray();
                var nextNode = nextNodes.Length == 0 ? double.NaN : nextNodes[0];
                if (double.IsNaN(nextNode))
                    continueOrNot = false;
                else
                    CurrentNode = Nodes.First(n => n.NodeId == (int)nextNode);
            }

            return State.result;
        }

        public JObject SerializeState() => this.State.ConvertToJson();
    }
}